lib/ansible/modules/monitoring/zabbix/zabbix_host.py

```diff
@@ -556,15 +556,16 @@ def check_all_properties(self, host_id, group_ids, status, interfaces, template_
                 return True
 
         if inventory_mode:
-            if LooseVersion(self._zbx_api_version) <= LooseVersion('4.4.0'):
-                if host['inventory']:
+            if host['inventory']:
+                # The way to read the inventory_mode is different on versions between [4.0.0, 4.4.1)
+                if LooseVersion(self._zbx_api_version) >= LooseVersion('4.0.0') and LooseVersion(self._zbx_api_version) < LooseVersion('4.4.1'):
```

Comment:
The following code should processing is Zabbix version 4.4.0 or less.

```
                    if int(host['inventory']['inventory_mode']) != self.inventory_mode_numeric(inventory_mode):
                        return True
```

But Zabbix version 3.0/3.2/3.4 isn't included following condition, so error occurs that when parameter updating.  
(Zabbix 3.0 is still in [support period](https://www.zabbix.com/life_cycle_and_release_policy) and needs to be supported by this module)

```
                if LooseVersion(self._zbx_api_version) >= LooseVersion('4.0.0') and LooseVersion(self._zbx_api_version) < LooseVersion('4.4.1'):
```

Could you please reconsider the above condition?

Reply:
Sory but I still don´t get it. There are two conditions:

- `if LooseVersion(self._zbx_api_version) >= LooseVersion('4.0.0') and LooseVersion(self._zbx_api_version) < LooseVersion('4.4.1'):`
    In this case check for `host['inventory']['inventory_mode']`. This covers Zabbix [4.0.0, 4.4.1)
- `else`:
    In this case check for `host['inventory_mode']`. This covers Zabbix (0.0.0, 4.0.0) and [4.4.1, 4.4.1+)

So as far I can tell your concern is about supporting version 3.0/3.2/3.4 which are covered by the `else` section because on those versions we need to check for `host['inventory_mode']` and not `host['inventory']['inventory_mode']`.

Is there something I'm missing (or maybe on the translation, English is not my first language, it is Spanish).

Reply:
Could you deploy Zabbix 3.0 in your test environment and run this [test playbook](https://gist.github.com/sky-joker/1c6f20f108000fa4bd236969d1c8e907) on Zabbix 3.0?  
Then, the following error occurs.

```
(venv) [root@localhost ~]# ansible-playbook pr_test_66860_for_zabbix.yml -vvv
(snip)
TASK [update inventory_zabbix param for host] **********************************************************************************************************************************
task path: /root/dev/PR/66860/test.yml:53
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482 `" && echo ansible-tmp-1582966175.1346583-161015819237482="` echo /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482 `" ) && sleep 0'
Using module file /root/dev/PR/66860/ansible-1/lib/ansible/modules/monitoring/zabbix/zabbix_host.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-18759uhn0vly3/tmp83nxyjpl TO /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/ /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/root/dev/PR/66860/ansible-1/venv/bin/python /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py", line 102, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.monitoring.zabbix.zabbix_host', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py", line 925, in <module>
  File "/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py", line 885, in main
  File "/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py", line 565, in check_all_properties
KeyError: 'inventory_mode'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1582966175.1346583-161015819237482/AnsiballZ_zabbix_host.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible.modules.monitoring.zabbix.zabbix_host', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py\", line 925, in <module>\n  File \"/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py\", line 885, in main\n  File \"/tmp/ansible_zabbix_host_payload_vq80i3em/ansible_zabbix_host_payload.zip/ansible/modules/monitoring/zabbix/zabbix_host.py\", line 565, in check_all_properties\nKeyError: 'inventory_mode'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

**Zabbix 3.0 install procedure**

https://www.zabbix.com/jp/download?zabbix=3.0&os_distribution=centos&os_version=7&db=mysql&ws=apache


Thank you for in advance.

Reply:
I don't have an environment to test it, but I see that you already test it against all supported versions so I changed the test.

Reply:
Thank you @tikoflano  
I checked each Zabbix version source code for the host data JSON structure.  
Source code the following.

* [3.0](https://github.com/zabbix/zabbix/blob/aa382ebff0fe2a7da2d92f72bbc861327d6b6e3c/frontends/php/host_prototypes.php#L276
)
* [4.0](https://github.com/zabbix/zabbix/blob/98553278ba431b8d574187d96d4fad6c4dd4717b/frontends/php/host_prototypes.php#L287
)
* [4.2](https://github.com/zabbix/zabbix/blob/acdcb4e70e30fdd96d6d1c741c2b061c7b60c222/frontends/php/host_prototypes.php#L287
)
* [4.4](https://github.com/zabbix/zabbix/blob/d35544490505cc64371f504f429abdc46039a044/frontends/php/host_prototypes.php#L277
)

Looking at the source code, you can see that the JSON structure for return value changed from 4.4.  
I want to merge this PR, but devel branch is freeze, so merging is blocked now.  
see: https://github.com/ansible-collections/overview/blob/master/README.rst#timeline

I think probably the Zabbix module moves to `community.general`, but I don't know when `community.general` is created.  
If this PR to merge that I think you need to reopen this PR there, after when `community.general` created.